src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+ * accesses are strided, i.e. the addresses difference between consecutive
+ * accesses to B is constant. Array A's accesses, on the other hand, need not
+ * be strided. The access pattern for array A depends on the values stored in
+ * the array B. Moreover, to compute the address for array A's accesses, IMP
+ * needs to read the values of array B. The accesses to array A follow the
+ * following equation
+ *      address = base_address + (coefficient * value)
+ * where
+ *      address represents the access address of array A
+ *      base_address represents the base address of array A, i.e. &A[0]
+ *      coefficient represents the size of elements of A
+ *      value is the index into array A, i.e. the value B[i]
+ *
+ * Since in most cases the coefficient is a power of 2, we can replace it with
+ * a left arithmetic shift operation as follows:
+ *      address = base_address + (value << shift)
+ *
+ * Therefore, to calculate the prefetch address for array A, i.e. address, we
+ * need to calculate the base_address and shift. The IMP algorithm performs
+ * this calculation by solving the above equation using two pairs of
+ * (value, address). Thus, the problem now gets reduced to solving a system of
+ * linear equations with two variables. The solution to this linear system is
+ * the predicted pair (base_address, shift).
+ *
+ * Once the (base_address, shift) pair is known, then IMP can compute the
+ * predicted access to array A, i.e. address, whenever it observes a load to
+ * the indexing array B, i.e. value. Note that for this algorithm to work IMP
+ * must figure out the value of index, i.e. the value loaded from array B.
+ * Therefore, IMP must track access to array B and then uses these accesses,
+ * along with access to array A, to compute the indirect access parameters
+ * (base_address, shift). IMP uses a stream detector to track accesses to array
+ * B. For computing the indirect access parameters, IMP uses a structure called
+ * the IndirectPatternDetector (IPD). After IMP's IPD detects an indirect
+ * pattern, i.e. the once it is able to compute the indirect parameters, IMP
+ * stores the computed pair in its IndirectTable
+ *
+ * For more details about the IMP algorithm, please refer to the paper
+ * mentioned in the references.
+ *
+ * What does this implementation DOESN'T do:
+ * There are some differences between this implementation and the algorithm
+ * described in the paper. Namely, this implementation does not support
+ * indirect prefetching for multi-way and multi-level accesses as described in
+ * section 3.3.2 of the paper
+ */
+
+#ifndef __MEM_CACHE_IMP_HH__
+#define __MEM_CACHE_IMP_HH__
+
+#include <map>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "mem/cache/prefetch/queued.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/IMPv2Prefetcher.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+    namespace prefetch
+    {
+        namespace IMPv2Internals
```

Comment:
The reason why I have kept these outside the IMP class is because these functions are not specific to just `IMP`. These are general functions which are used to modify saturating counters which are used in various structures of IMP. Making these functions a part of the IMP class gives the false impression of these functions being associated with an IMP object, which they are not. These are just helper functions.